src/components/base/Heading/index.tsx

```diff
@@ -2,26 +2,30 @@ import clsx from 'clsx';
 
 import { styles } from './styles.css';
 
-import type { FC, ReactNode } from 'react';
+import type { ComponentPropsWithoutRef, FC, ReactNode } from 'react';
 
-type Props = {
+type Tag = 'h1' | 'h2';
+
+type BaseProps<T extends Tag> = {
   /**
    * 見出しのタグ
    */
-  tag?: 'h1' | 'h2';
+  tag?: T;
   /**
    * 見出しのコンテンツ
    */
   children: ReactNode;
 };
 
-const Heading: FC<Props> = ({ tag = 'h1', children }) => {
-  switch (tag) {
-    case 'h1':
-      return <h1 className={clsx(styles.common, styles.h1)}>{children}</h1>;
-    case 'h2':
-      return <h2 className={clsx(styles.common, styles.h2)}>{children}</h2>;
-  }
+type Props<T extends Tag> = BaseProps<T> & Omit<ComponentPropsWithoutRef<T>, keyof BaseProps<T>>;
+
+const Heading: FC<Props<Tag>> = ({ tag = 'h1', children, className, ...rest }) => {
+  const Component = tag;
```

Comment:
ここで代入してるのってなんで？

Reply:
大文字始まりじゃないとタグとして識別してくれないからですね...
一応↓みたいに`tag: Tag = 'h1'`とするやり方もあるけど、このあたりは好みかな🤔

```suggestion
const Heading: FC<Props<Tag>> = ({ tag: Tag = 'h1', children, className, ...rest }) => {
  return (
    <Component className={clsx(styles.common, styles[Tag], className)} {...rest}>
      {children}
    </Component>
  );
```

Reply:
tagに型を指定してたらいけるってこと？

Reply:
すまん。若干↑のコード間違ってたので修正した🙏
Tag型を指定しているのではなく、tagというプロパティ名をTagという大文字始まりに変更している感じです🙇‍♂️

Reply:
あれ？ `tag: Tag = 'h1'` っていうのは、 `tag` が `Tag` 型であることを明示しているのではなくて、 `tag` というプロパティ名を `Tag` というプロパティ名に置き換えているということ:thinking-face:

Reply:
ですね！
JSの分割代入の時に`a: a1`みたいに名前変更できるので🙏
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

Reply:
そんなんあったんだ。しらなかった。そっちのほうがスマートそう

Reply:
@tyokinuhata 

88195fb こちらで対応しました🫡

Reply:
代入の形で表現すると「計算していないのになんで？」ってなりそうだけど、分割代入でプロパティ名を書き換える方法だと「名前を変えているんだな」と読み取れるので